File: app/src/main/java/com/gdgomaha/githubbrowser/home/list/RepoListPresenter.java
package com.gdgomaha.githubbrowser.home.list;

import com.gdgomaha.githubbrowser.data.RepoRepository;
import com.gdgomaha.githubbrowser.di.ForScreen;
import com.gdgomaha.githubbrowser.di.ScreenScope;
import com.gdgomaha.githubbrowser.home.Navigator;
import com.gdgomaha.githubbrowser.model.Repo;
import com.gdgomaha.githubbrowser.utils.DisposableManager;

import javax.inject.Inject;

import io.reactivex.android.schedulers.AndroidSchedulers;

@ScreenScope public final class RepoListPresenter {

    private final Navigator navigator;

    @Inject RepoListPresenter(
            Navigator navigator,
            @ForScreen DisposableManager disposableManager,
            RepoRepository repoRepository,
            RepoListViewModel viewModel) {
        this.navigator = navigator;
        disposableManager.add(repoRepository.getTrendingRepos()
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(viewModel::processRepos));
    }

    void repoSelected(Repo repo) {
        navigator.goToDetails(repo);
    }
}

File: app/src/main/java/com/gdgomaha/githubbrowser/home/list/RepoListPresenter.kt
package com.gdgomaha.githubbrowser.home.list

import com.gdgomaha.githubbrowser.data.RepoRepository
import com.gdgomaha.githubbrowser.di.ForScreen
import com.gdgomaha.githubbrowser.di.ScreenScope
import com.gdgomaha.githubbrowser.home.Navigator
import com.gdgomaha.githubbrowser.model.Repo
import com.gdgomaha.githubbrowser.utils.DisposableManager
import io.reactivex.android.schedulers.AndroidSchedulers
import javax.inject.Inject

@ScreenScope class RepoListPresenter @Inject constructor(
        private val navigator: Navigator,
        @ForScreen disposableManager: DisposableManager,
        repoRepository: RepoRepository,
        viewModel: RepoListViewModel) {

    init {
        disposableManager.add(repoRepository.trendingRepos
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ repos, error -> viewModel.processRepos(repos, error) }))
    }

    val selectedListener = fun (repo: Repo) {
        navigator.goToDetails(repo)
    }
}
